tests/components/imap/conftest.py

```diff
@@ -12,3 +16,135 @@ def mock_setup_entry() -> Generator[AsyncMock, None, None]:
         "homeassistant.components.imap.async_setup_entry", return_value=True
     ) as mock_setup_entry:
         yield mock_setup_entry
+
+
+@pytest.fixture
+def imap_capabilities() -> Generator[set[str], None]:
+    """Fixture to set the imap capabilities."""
+    return {"IDLE"}
+
+
+@pytest.fixture
+def imap_login_state() -> Generator[str, None]:
+    """Fixture to set the imap state after login."""
+    return AUTH
+
+
+@pytest.fixture
+def imap_select_state() -> Generator[str, None]:
+    """Fixture to set the imap capabilities."""
+    return SELECTED
+
+
+@pytest.fixture
+def imap_search() -> Generator[tuple[str, list[bytes]], None]:
+    """Fixture to set the imap search response."""
+    return EMPTY_SEARCH_RESPONSE
+
+
+@pytest.fixture
+def imap_fetch() -> Generator[tuple[str, list[bytes | bytearray]], None]:
+    """Fixture to set the imap fetch response."""
+    return TEST_FETCH_RESPONSE_TEXT_PLAIN
+
+
+@pytest.fixture
+def imap_pending_idle() -> Generator[bool, None]:
+    """Fixture to set the imap pending idle feature."""
+    return True
+
+
+@pytest.fixture
+def imap_close() -> Generator[AsyncMock, None]:
+    """Fixture to mock a side_effect the imap client close method."""
+    return AsyncMock()
+
+
+@pytest.fixture
+def imap_wait_server_push() -> Generator[AsyncMock, None]:
+    """Fixture to mock the imap client wait_server_push method."""
+    return AsyncMock()
+
+
+@pytest.fixture
+async def mock_imap_protocol(
+    imap_search: tuple[str, list[bytes]],
+    imap_fetch: tuple[str, list[bytes | bytearray]],
+    imap_capabilities: set[str],
+    imap_pending_idle: bool,
+    imap_login_state: str,
+    imap_select_state: str,
+    imap_wait_server_push: AsyncMock,
+    imap_close: AsyncMock,
+) -> Generator[MagicMock, None]:
+    """Mock the aioimaplib IMAP protocol handler."""
+
+    class IMAP4ClientMock:
```

Comment:
Can we autospec the target instead when patching the client so we don't need to implement all these client details in the fixture here?

Reply:
The protocol state needs to be mocked (login, select, close), not sure how this can be done using autospec.

Reply:
Just set `side_effect` or `return_value` of those attributes on the patched and specced client mock.

Reply:
I am struggeling with create_autospec to get that work. It gives me `E       AttributeError: __enter__`, but I do not know why. If I test separate it works, but not in the context of the fixture.

This fails...
```python
with create_autospec(
        homeassistant.components.imap.coordinator.IMAP4_SSL
    ) as imap_mock:
```

Reply:
You can pass `autospec=True` to `patch` too.

Reply:
Also note that it will be the return value of the mock class that has the attributes. Ie, the instance of the class.